### PR TITLE
Fix client idle timeout ending sessions too early

### PR DESCRIPTION
**Pull request by atburke**
This PR fixes a bug with client idle timeouts. Previously, each SSH channel had its own client activity tracker, which meant that the session would be closed if any of the channels reached the timeout.

For example, a client may open a channel for terminal I/O and one for subsystem requests. If subsystem requests are only made at the beginning of the session, the session will time out after the idle time passes, even if the terminal channel has been active. 

This is fixed by making the channels share an activity tracker.

### CODE REVIEW

File: lib/sshutils/ctx.go [hunk 1]
```diff
@@ -63,18 +65,39 @@ type ConnectionContext struct {
...
+	// clientLastActive records the last time there was activity from the client.
+	clientLastActive time.Time
+
+	clock clockwork.Clock
+}
+
+type ConnectionContextOption func(c *ConnectionContext)
+
+// SetConnectionContextClock sets the connection context's internal clock.
+func SetConnectionContextClock(clock clockwork.Clock) ConnectionContextOption {
```
Comment (zmb3) [inline lib/sshutils/ctx.go:78]: Am I missing something or is this never actually used?

**Review by r0mant — COMMENTED** [review]
@atburke I thought that client_idle_timeout is respected properly? Could you add more information to this PR description regarding scenarios when this is not the case please?

**Review by r0mant — APPROVED** [review]

File: lib/srv/regular/sshserver_test.go [hunk 2]
```diff
@@ -336,41 +336,78 @@ func TestInactivityTimeout(t *testing.T) {
...
+
+	t.Run("Normal timeout", func(t *testing.T) {
+		f := newCustomFixture(t, mutateCfg)
+
+		// If all goes well, the client will be closed by the time cleanup happens,
+		// so change the assertion on closing the client to expect it to fail
+		f.ssh.assertCltClose = require.Error
+		se, err := f.ssh.clt.NewSession(context.Background())
+		require.NoError(t, err)
+		defer se.Close()
```
Comment (jakule) [inline lib/srv/regular/sshserver_test.go:376]: nit:
```suggestion
		t.Cleanup(se.Close)
```

File: lib/srv/regular/sshserver_test.go [hunk 3]
```diff
@@ -336,41 +336,78 @@ func TestInactivityTimeout(t *testing.T) {
...
+
+	t.Run("Reset timeout on input", func(t *testing.T) {
+		f := newCustomFixture(t, mutateCfg)
+
+		// If all goes well, the client will be closed by the time cleanup happens,
+		// so change the assertion on closing the client to expect it to fail
+		f.ssh.assertCltClose = require.Error
+		se, err := f.ssh.clt.NewSession(context.Background())
+		require.NoError(t, err)
+		defer se.Close()
```
Comment (jakule) [inline lib/srv/regular/sshserver_test.go:388]:
```suggestion
		t.Cleanup(se.Close)
```

File: lib/srv/regular/sshserver_test.go [hunk 4]
```diff
@@ -336,41 +336,78 @@ func TestInactivityTimeout(t *testing.T) {
...
+
+		stdin, err := se.StdinPipe()
+		require.NoError(t, err)
+		t.Cleanup(func() { require.ErrorIs(t, stdin.Close(), io.EOF) })
+
+		endCh := make(chan error)
+		go func() { endCh <- f.ssh.clt.Wait() }()
```
Comment (jakule) [inline lib/srv/regular/sshserver_test.go:395]: I don't think this goroutine ever ends. If the test passes, this write will block forever, and this goroutine will never exit.

File: lib/srv/regular/sshserver_test.go [hunk 5]
```diff
@@ -336,41 +336,78 @@ func TestInactivityTimeout(t *testing.T) {
...
+	waitForTimeout := func(t *testing.T, f *sshTestFixture, se *tracessh.Session) {
+		stderr, err := se.StderrPipe()
+		require.NoError(t, err)
+		stdErrCh := startReadAll(stderr)
 
-	se, err := f.ssh.clt.NewSession(context.Background())
-	require.NoError(t, err)
-	defer se.Close()
+		endCh := make(chan error)
+		go func() { endCh <- f.ssh.clt.Wait() }()
```
Comment (jakule) [inline lib/srv/regular/sshserver_test.go:346]: I don't think this goroutine ever ends. If the test passes, this write will block forever, and this goroutine will never exit.

**Review by jakule — APPROVED** [review]
I left a few comments.

**Pull request merged**